src/NetAnalyzers/Core/Microsoft.CodeQuality.Analyzers/ApiDesignGuidelines/ConsiderPassingBaseTypesAsParameters.cs

```diff
@@ -0,0 +1,227 @@
+                MicrosoftCodeQualityAnalyzersResources.ResourceManager,
+                typeof(MicrosoftCodeQualityAnalyzersResources));
+
+        private static readonly LocalizableString s_localizableDescription =
+            new LocalizableResourceString(
+                nameof(MicrosoftCodeQualityAnalyzersResources.ConsiderPassingBaseTypesAsParametersDescription),
+                MicrosoftCodeQualityAnalyzersResources.ResourceManager,
+                typeof(MicrosoftCodeQualityAnalyzersResources));
+
+        internal static readonly DiagnosticDescriptor Rule =
+            DiagnosticDescriptorHelper.Create(
+                RuleId,
+                s_localizableTitle,
+                s_localizableStandardMessage,
+                DiagnosticCategory.Design,
+                RuleLevel.IdeHidden_BulkConfigurable,
+                description: s_localizableDescription,
+                isPortedFxCopRule: true,
+                isDataflowRule: false);
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(Rule);
+
+        public override void Initialize(AnalysisContext context)
+        {
+            context.EnableConcurrentExecution();
+            context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.None);
+
+            context.RegisterCompilationStartAction(context =>
+            {
+                var eventArgsTypeSymbol = context.Compilation.GetOrCreateTypeByMetadataName(WellKnownTypeNames.SystemEventArgs);
+
+                context.RegisterOperationBlockStartAction(context =>
+                {
+                    if (context.OwningSymbol is not IMethodSymbol methodSymbol ||
+                        methodSymbol.IsOverride ||
+                        methodSymbol.Parameters.Length == 0 ||
+                        methodSymbol.IsImplementationOfAnyInterfaceMember() ||
+                        methodSymbol.HasEventHandlerSignature(eventArgsTypeSymbol) ||
+                        !context.Options.MatchesConfiguredVisibility(Rule, methodSymbol, context.Compilation,
+                            defaultRequiredVisibility: SymbolVisibilityGroup.All) ||
+                        context.Options.IsConfiguredToSkipAnalysis(Rule, methodSymbol,
+                            context.Compilation))
+                    {
+                        return;
+                    }
+
+                    var typesUsedPerParameter = PooledConcurrentDictionary<IParameterSymbol, PooledConcurrentSet<ITypeSymbol>>.GetInstance(SymbolEqualityComparer.Default);
+                    foreach (var param in methodSymbol.Parameters)
+                    {
+                        typesUsedPerParameter.AddOrUpdate(param, PooledConcurrentSet<ITypeSymbol>.GetInstance(SymbolEqualityComparer.Default), (param, value) => value);
+                    }
+
+                    context.RegisterOperationAction(context =>
```

Comment:
I could have registered multiple operations and checked that the target was of type `IParameterReferenceOperation` but it felt easier this way.